# App's profitable for the google play markets and App store

This is a project that involves looking through apps in the google play market and the App store in Apple. We put oureselves in the role of a Data Analyst working in a company that develops apps that are free to download and install. We are tasked with finding what kind of apps have the most ratings and are generally well-liked by people and transfer that knowledge to the developers.

## Opening and Exploring the Data

There are 2 million apps each for google play and the app store. Analysing all those apps will take years. Luckily we can sample out and extract two datasets which seem to provide the relevent information we can use to analyse the data belonging to each app on a small scale (about a few thousand values or so). 

We can download the csv files for the android and apple datasets here:

[android dataset](https://www.kaggle.com/lava18/google-play-store-apps)

[apple dataset](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps)


We start by opening and reading through both the apple store and google play store datasets

In [1]:
from csv import reader

# The ios dataset
open_file = open('/Users/Tejas/csv_files/AppleStore.csv')
read_file = reader(open_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]

# The android dataset
open_file = open('/Users/Tejas/csv_files/googleplaystore.csv')
read_file = reader(open_file)
android = list(read_file)
android_header = android[0]
android = android[1:]




Next, we define a function that lets us explore through the data, figure out the shape of the data and what it contains. Our function takes in four parameters for slicing the data and for printing out the rows and columns to see the shape.

In [2]:
def explore_data(dataset, start, end, row_and_column=False):
    data_slice = dataset[start:end]
    for row in data_slice:
        print(row)
        print('\n')
        
    if row_and_column:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
        
print(android_header)
print('\n')
print(explore_data(android, 0, 2, True))
print('\n')
print(ios_header)
print('\n')
print(explore_data(ios, 0, 2, True))

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13
None


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '21

## Deleting Wrong Information

The Google Play data set has a dedicated discussion section, and we can see that one of the discussions outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.


In [3]:
print(android[10472]) # incorrect row
print('\n')
print(android_header) # header for android dataset
print('\n')
print(android[5]) # correct row

['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Paper flowers instructions', 'ART_AND_DESIGN', '4.4', '167', '5.6M', '50,000+', 'Free', '0', 'Everyone', 'Art & Design', 'March 26, 2017', '1.0', '2.3 and up']


We can see that the value for the category column is missing in row number 10472. We can remove this row for now, but know that there are other, better ways to correct missing data present in rows like these.

In [4]:
print(len(android)) # number of rows in the dataset before we removed the incorrect row
del(android[10472])
print(len(android)) # number of rows after

10841
10840


## Removing Duplicate Entries

### Part One

We can see that the apps in the google play store have more than one entry.

In [5]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)  
        print('\n')

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']




Working with duplicate entries in data sets like these is usually deprecated. We need to find a way to remove all of the duplicate entries present in the datasets. Fortunately, we know of a method that can help us. 

We create two lists, one for the duplicates and one for the unique entries. We append the values corresponding to those lists respectively.

In [6]:
duplicates = []
uniques = []

for app in android:
    name = app[0]
    if name in uniques:
        duplicates.append(name)
    else:
        uniques.append(name)
        
print('Examples of duplicate entries:', duplicates[:15])
print('\n')
print('Number of duplicate entries:', len(duplicates))
        

Examples of duplicate entries: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


Number of duplicate entries: 1181


From this we can see that the total number of duplicate entries is `1181`. 

One way to remove rows like these would be to remove them randomly. But we can think of a better alternative :-

We take a look at the reviews column present in the android data. The numbers represnt the number of ratings an app has received over a giver period of time. The higher the reviews, the higher the rating, and the more recently the data has been collected from that app. We can use this knowledge to filter out those rows with the highest number of reviews and only allow one entry per row to be taken in for analysis. 

This is how we do it:

### Part Two

Lets start by creating a dictionary called reviews_max. The keys will be the app names and the values will be the highest amount of reviews for that app. We only allow one entry per row.

In [7]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name not in reviews_max:
        reviews_max[name] = n_reviews


In the previous code cell, we found that there are 1,181 cases where an app occurs more than once, so the length of our dictionary (of unique apps) should be equal to the difference between the length of our data set and 1,181.

In [8]:
print(len(reviews_max) )
print(len(android)- 1181)

9659
9659


It seems that our expected and actual values match. Therfore we can proceed to the next step.

We first start by creating two lists : `android_clean` and `already_added`. We loop over the android dataset, isolating the name and the number of reviews of the app. 

We write a conditional statement inside the loop that specifies if the name is not in `already_added` and if there are no more than one entry of apps with the highest number of reviews present, then append the row(app) to the `android_clean` list and the name to the `already_added` list. 

For example, the app **"Box"** has three entries where the max number of reviews is the same. We can't have that present in our list, so we need to add that supplementary condition.


In [9]:
android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews ) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)


Now lets explore our dataset and confirm that the number of rows is 9,659

In [10]:
explore_data(android_clean, 0, 2, True) # It is 9,659

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 9659
Number of columns: 13


## Removing Non-English Apps

### Part One

In our dataset, there are entries that are written in non english text. We need to figure out a way to remove thos entries from our dataset. We start by defining a function which takes a string as a parameter and checks the ordinal value of the character present in that string. 

According to ASCII, the non english ordinal values are greater than 127. so we supply that condition into our function as follows:

In [11]:
def is_english(a_string):
    for char in a_string:
        if ord(char) > 127:
            return False
    return True

print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Instachat 😂'))
print(is_english('Docs To Go™ Free Office Suite'))

True
False
False
False


### Part Two

We can see that our function works for those strings that contain non english characters. But it also filters out characters like emojis and special characters like the trademark symbol. We need to make sure that those special characters are included in our string, because they belong to apps that include english names. We insert some conditions into our function as follows:

In [12]:
def is_english(a_string):
    non_ascii = 0
    
    for char in a_string:
        if ord(char) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))
            

True
True


The function is not perfect, and we will learn better ways to filter out characters like these over time. But for now, this method will suffice.

In [13]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)
        
for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)

In [14]:
explore_data(android_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up']


Number of rows: 9614
Number of columns: 13


In [15]:
explore_data(ios_english, 0, 3, True)

['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 6183
Number of columns: 16


Note that we are left with 10795 and 6183 rows for both datasets.

## Isolating the Free Apps

As mentioned previously, our developers only build apps that are free to download and install. So we have to isolate the free, english apps that are present in our datasets. We do this by specifying the price column values in each dataset corresponds to either 0 or 0.0

In [16]:
android_final = []
ios_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)
        
        
print(len(android_final))
print(len(ios_final))

8864
3222


We're left with 8864 Android apps and 3222 iOS apps, which should be enough for our analysis.

## Most Common Apps by Genre

### Part one

We mentioned earlier that our apps revenue is generated by the number of users that install and use our app, since our revenue mostly comes from in-game adds. Our strategy is comprised of three steps:

1. Build a minimal version of the app and add it to the play store.
2. If the app does well, improve it further.
3. If there is a large number of sales and revenue generated from the app, then we make more money by adding it to the app store.

Let's start by creating a frequency table for the prime_genre column in the android dataset, and the Genre and Category column in the ios dataset so we can see what kind of apps people like the most and how much revenue they generate.

### Part Two

We'll build two functions we can use to analyze the frequency tables:

One function to generate frequency tables that show percentages
Another function that we can use to display the percentages in a descending order

In [17]:
def freq_table(dataset, index):
    total = 0
    table = {}
    
    for row in dataset:
        total += 1
        value = row[index]
        
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
            
    table_percentages = {}
    for key in table:
        percentage = (total / table[key]) * 100
        table_percentages[key] = percentage
        
    return table_percentages
    

In [18]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_list = []
    
    for key in table:
        key_vals_as_tuples = (table[key], key)
        table_list.append(key_vals_as_tuples)
        
    sorted_list = sorted(table_list, reverse=True)
    for entry in sorted_list:
        print(entry[1], ':', entry[0])
    

In [19]:
print(android_header)
print(ios_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


In [20]:
display_table(android_final, 1) # display table for the Category column

BEAUTY : 16724.528301886792
COMICS : 16116.363636363636
ART_AND_DESIGN : 15550.877192982454
PARENTING : 15282.758620689654
EVENTS : 14069.84126984127
WEATHER : 12484.507042253521
HOUSE_AND_HOME : 12142.465753424658
AUTO_AND_VEHICLES : 10809.756097560976
LIBRARIES_AND_DEMO : 10679.518072289156
ENTERTAINMENT : 10428.235294117647
EDUCATION : 8605.825242718447
FOOD_AND_DRINK : 8058.181818181818
MAPS_AND_NAVIGATION : 7148.387096774193
VIDEO_PLAYERS : 5574.842767295598
DATING : 5372.121212121212
BOOKS_AND_REFERENCE : 4665.263157894738
SHOPPING : 4454.271356783919
TRAVEL_AND_LOCAL : 4282.125603864734
SOCIAL : 3755.9322033898306
NEWS_AND_MAGAZINES : 3574.1935483870966
PHOTOGRAPHY : 3396.168582375479
HEALTH_AND_FITNESS : 3246.8864468864467
COMMUNICATION : 3088.501742160279
PERSONALIZATION : 3014.965986394558
SPORTS : 2944.8504983388707
MEDICAL : 2831.9488817891374
FINANCE : 2702.439024390244
PRODUCTIVITY : 2569.2753623188405
LIFESTYLE : 2561.849710982659
BUSINESS : 2177.886977886978
TOOLS : 118

In [21]:
display_table(ios_final, -5) # display table for the prime_genre column

Catalogs : 80550.0
Navigation : 53700.0
Medical : 53700.0
Book : 23014.285714285714
Business : 18952.941176470587
Reference : 17900.0
Food & Drink : 12392.307692307691
Weather : 11507.142857142857
Finance : 8950.0
Travel : 8055.0
News : 7493.023255813953
Lifestyle : 6317.64705882353
Productivity : 5753.571428571428
Health & Fitness : 4956.923076923077
Music : 4881.818181818182
Sports : 4669.565217391305
Utilities : 3977.777777777778
Shopping : 3835.7142857142853
Social Networking : 3039.622641509434
Education : 2730.5084745762715
Photo & Video : 2013.75
Entertainment : 1268.5039370078741
Games : 171.93169690501603


In [22]:
display_table(android_final, -4) # display table for the genre column

Video Players & Editors;Creativity : 886400.0
Trivia;Education : 886400.0
Travel & Local;Action & Adventure : 886400.0
Tools;Education : 886400.0
Strategy;Education : 886400.0
Strategy;Creativity : 886400.0
Strategy;Action & Adventure : 886400.0
Simulation;Education : 886400.0
Role Playing;Brain Games : 886400.0
Racing;Pretend Play : 886400.0
Puzzle;Education : 886400.0
Parenting;Brain Games : 886400.0
Music & Audio;Music & Video : 886400.0
Lifestyle;Pretend Play : 886400.0
Lifestyle;Education : 886400.0
Health & Fitness;Education : 886400.0
Health & Fitness;Action & Adventure : 886400.0
Entertainment;Education : 886400.0
Communication;Creativity : 886400.0
Comics;Creativity : 886400.0
Casual;Music & Video : 886400.0
Card;Action & Adventure : 886400.0
Books & Reference;Education : 886400.0
Art & Design;Pretend Play : 886400.0
Art & Design;Action & Adventure : 886400.0
Arcade;Pretend Play : 886400.0
Adventure;Education : 886400.0
Video Players & Editors;Music & Video : 443200.0
Sports;A

## Most Popular Apps by Genre on the App Store

We use our frequency table function to generate a table for the genres column in the ios dataset. We use a nested for loop to iterate over the genres column and find the average number of ratings that column has, to find out which of those apps are the most popular.

In [23]:
genres_ios = freq_table(ios_final, -5)

for genre in genres_ios:
    total = 0
    length = 0
    
    for app in ios_final:
        genres_app = app[-5]
        
        if genres_app == genre:
            n_rating = float(app[5])
            total += n_rating
            length += 1
    
    avg_ratings = total / length

    print(genre, ':', avg_ratings)
            

Social Networking : 71548.34905660378
Photo & Video : 28441.54375
Games : 22788.6696905016
Music : 57326.530303030304
Reference : 74942.11111111111
Health & Fitness : 23298.015384615384
Weather : 52279.892857142855
Utilities : 18684.456790123455
Travel : 28243.8
Shopping : 26919.690476190477
News : 21248.023255813954
Navigation : 86090.33333333333
Lifestyle : 16485.764705882353
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Book : 39758.5
Finance : 31467.944444444445
Education : 7003.983050847458
Productivity : 21028.410714285714
Business : 7491.117647058823
Catalogs : 4004.0
Medical : 612.0


## Most Popular Apps by Category on the Google Play Market

We follow the same principle we used while looping over the genres column in the app store dataset.

In [24]:
categories_android = freq_table(android_final, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:            
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

ART_AND_DESIGN : 1986335.0877192982
AUTO_AND_VEHICLES : 647317.8170731707
BEAUTY : 513151.88679245283
BOOKS_AND_REFERENCE : 8767811.894736841
BUSINESS : 1712290.1474201474
COMICS : 817657.2727272727
COMMUNICATION : 38456119.167247385
DATING : 854028.8303030303
EDUCATION : 1833495.145631068
ENTERTAINMENT : 11640705.88235294
EVENTS : 253542.22222222222
FINANCE : 1387692.475609756
FOOD_AND_DRINK : 1924897.7363636363
HEALTH_AND_FITNESS : 4188821.9853479853
HOUSE_AND_HOME : 1331540.5616438356
LIBRARIES_AND_DEMO : 638503.734939759
LIFESTYLE : 1437816.2687861272
GAME : 15588015.603248259
FAMILY : 3695641.8198090694
MEDICAL : 120550.61980830671
SOCIAL : 23253652.127118643
SHOPPING : 7036877.311557789
PHOTOGRAPHY : 17840110.40229885
SPORTS : 3638640.1428571427
TRAVEL_AND_LOCAL : 13984077.710144928
TOOLS : 10801391.298666667
PERSONALIZATION : 5201482.6122448975
PRODUCTIVITY : 16787331.344927534
PARENTING : 542603.6206896552
WEATHER : 5074486.197183099
VIDEO_PLAYERS : 24727872.452830188
NEWS_AND_

## Conclusion

From this knowledge, we can find out what kinds of apps available on both platforms are the most popular by the average amount of ratings they have and how we can pass this knowledge onto the developers to create apps like these to boost sales.

Thank you for reading.